File: crates/l1-kernel-rs/src/schema.rs
//! Thread-safe string-event schema registry candidate for the L1 kernel.

use std::collections::BTreeMap;
use std::sync::{Arc, Mutex, MutexGuard, OnceLock, PoisonError};

use serde::{Deserialize, Serialize};

/// One owner-qualified string event description.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct EventSchema {
    /// Stable event name.
    pub name: String,
    /// Subsystem that owns the event name.
    pub owner: String,
    /// Human-readable event description.
    pub description: String,
}

/// Thread-safe sorted event-schema registry.
pub struct EventSchemaRegistry {
    events: Mutex<BTreeMap<String, EventSchema>>,
}

impl EventSchemaRegistry {
    /// Create an empty schema registry.
    pub fn new() -> Self {
        Self {
            events: Mutex::new(BTreeMap::new()),
        }
    }

    /// Register an event, rejecting a name claimed by another owner.
    pub fn register_event(
        &self,
        name: impl Into<String>,
        owner: impl Into<String>,
        description: impl Into<String>,
    ) -> bool {
        let name = name.into();
        let owner = owner.into();
        let mut events = self.lock_events();
        if let Some(existing) = events.get(&name)
            && existing.owner != owner
        {
            return false;
        }
        events.insert(
            name.clone(),
            EventSchema {
                name,
                owner,
                description: description.into(),
            },
        );
        true
    }

    /// Return whether an event name is registered.
    pub fn has_event(&self, name: &str) -> bool {
        self.lock_events().contains_key(name)
    }

    /// Return a deterministic name-sorted snapshot of the catalog.
    pub fn list_events(&self) -> Vec<EventSchema> {
        self.lock_events().values().cloned().collect()
    }

    /// Clear all registrations.
    pub fn reset(&self) {
        self.lock_events().clear();
    }

    fn lock_events(&self) -> MutexGuard<'_, BTreeMap<String, EventSchema>> {
        self.events.lock().unwrap_or_else(PoisonError::into_inner)
    }
}

impl Default for EventSchemaRegistry {
    fn default() -> Self {
        Self::new()
    }
}

static GLOBAL_SCHEMA: OnceLock<Mutex<Option<Arc<EventSchemaRegistry>>>> = OnceLock::new();

fn global_schema() -> &'static Mutex<Option<Arc<EventSchemaRegistry>>> {
    GLOBAL_SCHEMA.get_or_init(|| Mutex::new(None))
}

/// Return the process-wide event schema candidate.
pub fn get_schema() -> Arc<EventSchemaRegistry> {
    let mut slot = global_schema()
        .lock()
        .unwrap_or_else(PoisonError::into_inner);
    Arc::clone(slot.get_or_insert_with(|| Arc::new(EventSchemaRegistry::new())))
}

/// Register an event on the process-wide candidate registry.
pub fn register_event(
    name: impl Into<String>,
    owner: impl Into<String>,
    description: impl Into<String>,
) -> bool {
    get_schema().register_event(name, owner, description)
}

/// Check an event on the process-wide candidate registry.
pub fn has_event(name: &str) -> bool {
    get_schema().has_event(name)
}

/// List events from the process-wide candidate registry.
pub fn list_events() -> Vec<EventSchema> {
    get_schema().list_events()
}

/// Reset the process-wide candidate registry for test isolation.
pub fn reset_event_schema() {
    *global_schema()
        .lock()
        .unwrap_or_else(PoisonError::into_inner) = None;
}

File: systems/python-reference-runtime/l1/kernel/schema.py
"""Kernel string-event schema registry (W4.3).

SignalType (event.py) is the typed kernel bus surface; this module is the
catalog for STRING events emitted across L3 buses. Every event name must be
registered here with its owner before production use — a test enforces the
catalog is populated at boot. Ordering contract: FIFO within one channel,
no ordering guarantees across channels.
"""

from __future__ import annotations

import threading

_EVENTS: dict[str, dict] = {}
_LOCK = threading.Lock()


def register_event(name: str, owner: str, description: str = "") -> bool:
    """Register a string event name with its owning subsystem.

    Returns False when the name is already registered with a different owner.
    """
    with _LOCK:
        existing = _EVENTS.get(name)
        if existing and existing["owner"] != owner:
            return False
        _EVENTS[name] = {"owner": owner, "description": description}
        return True


def has_event(name: str) -> bool:
    """True when *name* is registered in the schema."""
    return name in _EVENTS


def list_events() -> list[dict]:
    """Registered events as a sorted list of {name, owner, description}."""
    with _LOCK:
        return [{"name": n, **e} for n, e in sorted(_EVENTS.items())]


def reset_event_schema() -> None:
    """Drop all registrations (tests / factory reset)."""
    global _EVENTS
    with _LOCK:
        _EVENTS = {}
